examples/misc/test/language_tour/operators_test.dart

```diff
@@ -159,13 +159,18 @@ void main() {
     assert((value & ~bitmask) == 0x20); // AND NOT
     assert((value | bitmask) == 0x2f); // OR
     assert((value ^ bitmask) == 0x2d); // XOR
+
     assert((value << 4) == 0x220); // Shift left
     assert((value >> 4) == 0x02); // Shift right
+
+    // Shift right example that results in different behavior on web
+    // due to negative operand:
```

Comment:
This is delicate, as always, but I think we need to get it right (such that we don't send any readers who actually need to use bitwise operations on the web in the wrong direction).

If the operand is negative and we're on the web then the 32 bit unsigned value which is used as the receiver (left hand operand) or right hand operand in a bitwise operation is obtained by masking off all higher bits of the 2's complement representation of the IEEE 754 encoding, which can yield every single one of the 2^32 bit vectors that an unsigned 32 bit int can be. So we don't know anything about the outcome of `.toSigned(32)`; for instance, it could be a negative number and it could be a positive number.

You could say that if the initial value is negative and "not too negative (-2^32 or bigger)" then masking the value to 32 bits and doing `.toSigned(32)` _will_ yield the same value. But then we'd need to mention "32 bits".

In particular, it really does matter that non-zero bits at position 32 or bigger are outright discarded before we even begin to compute the bitwise operator result, and this matters for every bitwise operator:

```dart
void main() {
  var big = 1024 * 1024 * 1024 * 1024; // 2^40, >32 bits, and OK on the web.
  print(big); // '1099511627776'.
  print(big >> 0); // '0'.
  print(big >>> 0); // '0'.
  print(-big >> 0); // '0'.
  print(-big >>> 0); // '0'.
}
```

I think this implies that it isn't very helpful to establish the rule of thumb in the minds of readers that "I just need to look out for negative operands to `>>`", we really need to mention the fact that these operations take place in a 32 bit unsigned representation on the web.

So how about this:
```suggestion
    // because the operand value changes when masked to 32 bits:
```

and then we should probably have a link to [Numbers in Dart](https://dart.dev/guides/language/numbers).

Reply:
Ah, we have that already.

Reply:
I really appreciate the thoughtful insights and the focus on what really matters for users who need to use bitwise operations on the web.

I like your adjusted text and it makes sense to me. I opened https://github.com/dart-lang/site-www/pull/5065 which applies these changes as this PR was accidentally merged without applying them.

Thanks again!

Reply:
Thanks! And thank you for the kind words!